lib/auth/github.go

```diff
@@ -84,6 +91,95 @@ func (a *Server) upsertGithubConnector(ctx context.Context, connector types.Gith
 	return nil
 }
 
+// HTTPRequester allows a net/http.Client to be mocked for tests.
+type HTTPRequester interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// checkGithubFeatureSupport returns an error if any of the Github
+// organizations specified in this connector use external SSO. SSO
+// is a Teleport Enterprise feature that should not be allowed for OSS.
+// If client is nil a net/http.Client will be used.
+func checkGithubFeatureSupport(conn types.GithubConnector, client HTTPRequester) error {
+	version := modules.GetModules().BuildType()
+	if version == modules.BuildEnterprise {
+		return nil
+	}
+
+	if client == nil {
+		transport, err := defaults.Transport()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		client = &http.Client{
+			Transport: transport,
+		}
+	}
+
+	checkOrg := func(org string) error {
+		// A Github organization will have a "sso" page reachable if it
+		// supports external SSO. There doesn't seem to be any way to get this
+		// information from the Github REST API without being an owner of the
+		// Github organization, so check if this exists instead.
+		ssoURL := fmt.Sprintf("%s/%s/%s", githubOrgsURL, url.PathEscape(org), "sso")
+
+		const retries int = 3
+		var resp *http.Response
+		for i := 0; i < retries; i++ {
+			ctx, cancel := context.WithTimeout(context.Background(), defaults.HTTPRequestTimeout)
+			defer cancel()
+
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, ssoURL, nil)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			var urlErr *url.Error
+			resp, err = client.Do(req)
```

Comment:
Does this noticeably slow down logins? I wonder if we should consider some sort of caching, e.g. cache results with TTL for a few minutes/hours as I don't imagine this is a setting that will change often.

Reply:
I think that we should cache for at least an hour. GH has rate limiting 5k requests per hour for the free plan and 15k for paid one https://docs.github.com/en/developers/apps/building-github-apps/rate-limits-for-github-apps
Maybe it doesn't sound like we could hit it, but I don't think that checking this setting too often makes sense.

Reply:
I used https://github.com/patrickmn/go-cache to do this.

Reply:
Updated to use our `utils.FnCache` package instead.